flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -61,6 +74,144 @@ static void transferAllStateDataToDirectory(
 		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
 	}
 
+	public static void uploadStateFiles(
+		CheckpointStreamFactory checkpointStreamFactory,
+		SnapshotDirectory localBackupDirectory,
+		Set<StateHandleID> baseSstFiles,
+		int uploadingThreadNum,
+		CloseableRegistry snapshotCloseableRegistry,
+		@Nonnull ConcurrentHashMap<StateHandleID, StreamStateHandle> sstFiles,
+		@Nonnull ConcurrentHashMap<StateHandleID, StreamStateHandle> miscFiles) throws Exception {
+
+		Preconditions.checkState(localBackupDirectory.exists());
+
+		FileStatus[] fileStatuses = localBackupDirectory.listStatus();
+		if (fileStatuses != null) {
+			ExecutorService executorService = createExecutorService(uploadingThreadNum);
+
+			try {
+				List<Runnable> runnables = createUploadRunnables(
```

Comment:
I think using `Callable<StreamStateHandle>` (path -> StreamStateHandle, as it was in `uploadLocalFileToCheckpointFs`) is simpler and stateless, comparing to using `Runnable` and implicit adding results into the concurrent map.

What if we implement `createUploadCallables` and use `CompletableFuture.callAsync` to create `Map<StateHandleID, CompletableFuture<StreamStateHandle>>`? Then the futures (map.values()) could be waited for and the map could be converted into the result `Map<StateHandleID, StreamStateHandle>`.

Reply:
Hi @azagrebin, when I want to implement this by using `Callable`, I can not find `java.util.concurrent.CompletableFuture.callAsync` function(JDK 8), after searching in flink project, I just find `org.apache.flink.runtime.rpc.MainThreadExecutable#callAsync`,  Did I miss something?

In other words, If I use `Callable` instead of `Runnable`, should I implement a 
 `ThrowingCallable` just as `ThrowingRunnable`.

If I use `Runnable` to implement this, how about create a `CompletableFuture<StreamStateHandle>` instance and pass to the called function `uploadLocalFileToCheckpointFs(Path, CompletableFuture)`, in `uploadLocalFileToCheckpointFs` we complete the passed in `CompletableFuture<StreamStateHandle>`,  such as below

```
Map<StateHandleID, CompletableFuture<StreamStateHandle>> futures = new HashMap();
// futures will update in createUploadRunnables
List<Runnable> runnables = createUploadRunnables(sstFilePaths, miscFilePaths, futures);
for (Runnable runnable : runnables) {
    CompletableFuture.runAsync(runnable, executorService);
}
FutureUtils.waitForAll(futures.values()).get();

....

private static StreamStateHandle uploadLocalFileToCheckpointFs(Path path, CompletableFuture future) {
   try {
       ...
       future.complete(...)
   } finally {
       ....
   }
}
```

Reply:
Hi @klion26 , I think this could work:
```
futures .put(stateHandleID , 
  CompletableFuture.supplyAsync(CheckedSupplier.unchecked(
    () -> uploadLocalFileToCheckpointFs(path)), executorService))
```
`CompletableFuture.supplyAsync` seems to do already internally what you suggest.
Also I would run parallel upload separately for  `sstFilePaths` and `miscFilePaths`, one after another, so that `RocksDbStateDataTransfer` does not need to know about file structure.

Reply:
@azagrebin thank you for your reply, I'll update the code according to your suggestions.